internal/ratelimit/handler.go

```diff
@@ -0,0 +1,125 @@
+	match := pathRegex.FindStringSubmatch(path)
+	for i, name := range pathRegex.SubexpNames() {
+		switch name {
+		case "resource":
+			res = match[i]
+			r, ok := resource.FromPlural(res)
+			if !ok {
+				res = resource.Unknown.String()
+				err = fmt.Errorf("unknown resource")
+			} else {
+				res = r.String()
+			}
+		case "action":
+			act = match[i]
+			if act != "" {
+				_, ok := action.Map[act]
+				if !ok {
+					act = action.Unknown.String()
+					err = fmt.Errorf("unknown action")
+				}
+			}
+		case "id":
+			id = match[i]
+		}
+	}
+
+	switch act {
+	case "":
+		switch id {
+		case "":
+			switch method {
+			case http.MethodGet:
+				act = action.List.String()
+			case http.MethodPost:
+				act = action.Create.String()
+			default:
+				act = action.Unknown.String()
+				err = fmt.Errorf("unknown action")
+			}
+		default:
+			switch method {
+			case http.MethodGet:
+				act = action.Read.String()
+			case http.MethodDelete:
+				act = action.Delete.String()
+			case http.MethodPut:
+				act = action.Update.String()
+			default:
+				act = action.Unknown.String()
+				err = fmt.Errorf("unknown action")
+			}
+		}
+	}
+	return
+}
+
+// Handler is an http middleware handler that checks if a request is allowed
+// using the provided rate limiter. If the request is allowed, the next handler
+// is called. Otherwise a 429 is returned with the Retry-After response header
+// set to the number of seconds the client should wait to make it's next request.
+func Handler(l *rate.Limiter, next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		path := req.URL.Path
+		method := req.Method
+
+		res, a, err := extractResourceAction(path, method)
+		if err != nil {
+			// TODO: what to do here?
+			// We could:
+			// 1. Bail early and block the request, but this prevents our existing
+			//    logic from running and presumably returning a 404.
+			// 2. Just skip doing rate limiting and call the next handler.
+			//    Since the controller *should* just 404, it is probably ok to allow
+			//    the request and not worry about limiting it.
+			// 3. Have a fall-back resource/action (i.e. resource.Unknown/action.Unknown)
+			//    and apply a rate limit to it just like any other resource+action pair.
+			//    This might be best, since it allows for limiting even invalid requests.
+			//    But we would need to document this Unknown case. If this is the case,
+			//    We should probably have extractResource always return a resource+action,
+			//    and not return an error.
+			rw.WriteHeader(http.StatusNotFound)
+			return
+		}
+
+		allowed, quota, err := l.Allow(res, a)
+		if err != nil && !errors.Is(err, rate.ErrLimiterFull) {
+			// TODO: what to do here?
+			// The only other error here should be rate.ErrLimitNotFound, which
+			// shouldn't be possible given how we initialize the limiter,
+			// so perhaps a 500 makes sense?
+			rw.WriteHeader(http.StatusInternalServerError)
```

Comment:
I could see a 503 when the error is `rate.ErrLimiterFull`. I was going to do a 429, but the 503 might be better to distinguish between being rate limited vs the server being overloaded.

However, in the case of `rate.ErrLimitNotFound` I don't think it would be temporary, since it would indicate that the limiter was not properly initialized.

Reply:
Hum... thinking a 429 if it's full and a 503 if it's not properly initialized.
Returning a 503 when the service is available but overloaded seems odd to me
since the client should likely behave the same way when experiencing a rate
limit or an overloaded service because it's out of quota.

Reply:
I am not sure I like a `503` for when it is not properly initialized, since it is not something that would be fixed by a configuration change, this would be a bug in the particular release and the only way the service would become available again is with an upgrade/downgrade. This is akin to a panic so I think a `500` would be better.

For when the limiter is full, I do kind of like `503` better. A `429` states:
> The 429 status code indicates that the user has sent too many requests in a given amount of time ("rate limiting").

But when the limiter is full, it is not that the user made too many requests, it is that the service is overwhelmed, which seems more in line with a `503` (my emphasis):
> The 503 (Service Unavailable) status code indicates that the server is currently unable to handle the request **due to a temporary overload** or scheduled maintenance, which will likely be alleviated after some delay.

In addition, with the 503 we can still set a `Retry-After` based on the time until quotas will be deleted, but we would not include the `RateLimit` header, since no quota was created.

As for clients, they could:
1. Minimally check for the `Retry-After` header and retry after that time.
2. Be explicit and check for either a 429 or 503 and the `Retry-After` header.

This would also help operators, since, with our existing monitoring, they could differentiate between requests that consumed too much quota (429), and requests that failed because of the limiter running out of space (503).

Reply:
44d49eb0d switches to a 503 on ErrLimiterFull, on other errors it is still a 500. @jimlambrt let me know if you disagree with my reasoning [above](https://github.com/hashicorp/boundary/pull/3825#discussion_r1376299682), I'm happy to continue discussing it.

Reply:
I'm fine with it.  My thinking was: few clients handle 503s as a retry-able error but most will deal with 429s properly.  In other words, a 503 will like result in the client returning an error to the caller and a 429 will likely be retried before returning an error to the caller.